pkg/validationfile/loader.go

```diff
@@ -52,124 +51,96 @@ func PopulateFromFiles(ds datastore.Datastore, filePaths []string) (*PopulatedVa
 // PopulateFromFilesContents populates the given datastore with the namespaces and tuples found in
 // the validation file(s) contents specified.
 func PopulateFromFilesContents(ds datastore.Datastore, filesContents map[string][]byte) (*PopulatedValidationFile, datastore.Revision, error) {
-	var revision datastore.Revision
-	var nsDefs []*core.NamespaceDefinition
+	var schema string
+	var objectDefs []*core.NamespaceDefinition
 	var caveatDefs []*core.CaveatDefinition
-	schema := ""
 	var tuples []*core.RelationTuple
+	var updates []*core.RelationTupleUpdate
+
+	var revision datastore.Revision
+
 	files := make([]ValidationFile, 0, len(filesContents))
 
+	// Parse each file into definitions and relationship updates.
 	for filePath, fileContents := range filesContents {
+		// Decode the validation file.
 		parsed, err := DecodeValidationFile(fileContents)
 		if err != nil {
 			return nil, datastore.NoRevision, fmt.Errorf("error when parsing config file %s: %w", filePath, err)
 		}
 
 		files = append(files, *parsed)
 
-		// Add schema-based namespace definitions.
+		// Disallow legacy sections.
+		if len(parsed.NamespaceConfigs) > 0 {
+			return nil, revision, fmt.Errorf("definitions must be specified in `schema`")
+		}
+
+		if len(parsed.ValidationTuples) > 0 {
+			return nil, revision, fmt.Errorf("relationships must be specified in `relationships`")
+		}
+
+		// Add schema definitions.
 		if parsed.Schema.CompiledSchema != nil {
 			defs := parsed.Schema.CompiledSchema.ObjectDefinitions
 			if len(defs) > 0 {
 				schema += parsed.Schema.Schema + "\n\n"
 			}
 
-			log.Info().Str("filePath", filePath).Int("schemaDefinitionCount", len(defs)).Msg("Loading schema definitions")
-			nsDefs = append(nsDefs, defs...)
+			log.Info().Str("filePath", filePath).Int("schemaDefinitionCount", len(parsed.Schema.CompiledSchema.OrderedDefinitions)).Msg("adding schema definitions")
+			objectDefs = append(objectDefs, defs...)
 			caveatDefs = append(caveatDefs, parsed.Schema.CompiledSchema.CaveatDefinitions...)
 		}
 
-		// Load the namespace configs.
-		log.Info().Str("filePath", filePath).Int("namespaceCount", len(parsed.NamespaceConfigs)).Msg("Loading namespaces")
-		for index, namespaceConfig := range parsed.NamespaceConfigs {
-			nsDef := core.NamespaceDefinition{}
-			nerr := prototext.Unmarshal([]byte(namespaceConfig), &nsDef)
-			if nerr != nil {
-				return nil, revision, fmt.Errorf("error when parsing namespace config #%v from file %s: %w", index, filePath, nerr)
-			}
-			nsDefs = append(nsDefs, &nsDef)
+		// Parse relationships for updates.
+		for _, rel := range parsed.Relationships.Relationships {
+			tpl := tuple.MustFromRelationship(rel)
+			updates = append(updates, tuple.Touch(tpl))
+			tuples = append(tuples, tpl)
 		}
```

Comment:
This seems like a nice snippet to refactor into a helper at `pkg/tuple/tuple.go`.